File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.py
from collections import deque

def solution(msg):
    answer = []
    d = {chr(x):x-64 for x in range(65, 91)}
    next_index = 27
    
    while True:
        if msg in d:
            answer.append(d[msg])
            break
            
        w = msg[0]
        for i in range(1, len(msg)):
            if w + msg[i] not in d:
                c = msg[i]
                answer.append(d[w])
                d[w+c] = next_index
                next_index += 1
                msg = msg[i:]
                break
            else:
                w += msg[i]
     
        
    return answer

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.js
function solution(msg) {
    var answer = [];
    
    // 1. 딕셔너리 초기화
    const dict = new Map();
    for (let i=0; i<26; i++) {
        dict.set(String.fromCharCode(65 + i), i+1);
    }
    
    // 2. 현재 입력과 일치하는 가장 긴 문자열 찾기
    let msgArray = msg.split('')
    let cur = ''
    while (msgArray.length > 0) {
        cur += msgArray.shift()
        if (dict.get(cur)) continue
        answer.push(dict.get(cur.slice(0, -1)))
        dict.set(cur, dict.size + 1)
        msgArray.unshift(cur.slice(-1))
        cur = ''
    }
    
    if (cur !== '') {
    answer.push(dict.get(cur));
    }
    
    return answer;
}